File: Generala-MF/FMTABLERO.vb
Public Class FMTABLERO

    'variable sobre la cual el formulario guarda el juego para operar
    'lo declaro con eventos
    Dim WithEvents _GeneralaBLL As BLL.GENERALA
    'lista de puntajes temporales posibles de un jugador dentro de la tirada para luego ser asignada
    Dim _tempPuntaje As List(Of Integer)

    'Carga el formulario del Tablero.
    Private Sub Tablero_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        'cargo el juego nuevo
        _GeneralaBLL = New BLL.GENERALA

        'tomo los usuarios de una nueva partida y los coloco en una lista de PLAYERS
        Dim PL As New List(Of BLL.PLAYER)
        For Each i In FMCONTENEDOR.liUSRPartida
            PL.Add(New BLL.PLAYER With {.usuarioBE = i})
        Next
        'le agrego la lista de players al juego
        _GeneralaBLL.agregarPlayers(PL)

        'pongo en 0 todos los datos de la GUI
        actualizarEstadoJuego()
        limpiaDados()
        cambiarEstadoBA(False)
        actualizarTempPuntaje(False)

    End Sub

#Region "Actualizar labels"
    'contiene todo lo que hay que refrescar por cada jugada
    Private Sub actualizarEstadoJuego()
        'genera el siguiente turno
        _GeneralaBLL.turnoNext()
        'muestra la tirada en curso
        LNumeroJugada.Text = _GeneralaBLL.generalaBE.rondas

        'actualiza label Jugador
        LTurnoJugador.Text = _GeneralaBLL.liPlayers.Item(_GeneralaBLL.generalaBE.turno).usuarioBE.usuario
        'actualiza el label de numero de tirada
        LNumeroTirada.Text = _GeneralaBLL.generalaBE.numeroJugada
        'actualiza la tabla de puntajes para cada jugador
        For i As Integer = 0 To (_GeneralaBLL.liPlayers.Count - 1)
            refreshScore(i)
        Next
    End Sub

    'se utiliza para realizar la actualizacion de los puntajes ya cargados.
    Private Sub refreshScore(jugador As Integer)
        Dim lb1 As String, lb2 As String
        'aca guardo el control a buscar
        Dim c As Control
        For i As Integer = 0 To 10
            lb1 = "SJ" & jugador + 1 & "A" & i
            For Each c In Me.Controls
                If c.Name = lb1 Then
                    c.Text = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorValor.Item(i)
                End If

            Next
        Next
        'armo un texto con el nombre del "label" que voy a buscar
        lb1 = "SJ" & jugador + 1 & "Resultado"
        lb2 = "LJugador" & jugador + 1
        'recorro todos los controles para encontrar el control del "label" que quiero buscar
        For Each c In Me.Controls
            'chequeo el nombre del "label" que busco
            If c.Name = lb1 Then
                'asigno el contenido de informacion que necesito al "label" encontrado
                c.Text = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.resultado
            End If
            If c.Name = lb2 Then
                c.Text = _GeneralaBLL.liPlayers.Item(jugador).usuarioBE.usuario
            End If

        Next
    End Sub

    'Actualiza los labels correspondientes al puntaje temporario
    Private Sub actualizarTempPuntaje(boo As Boolean)
        Dim lb As String
        Dim c As Control
        For i As Integer = 0 To 10
            lb = "LST" & i
            For Each c In Me.Controls
                If c.Name = lb Then
                    If boo Then
                        c.Text = _tempPuntaje.Item(i)
                    Else
                        c.Text = ""
                    End If
                End If
            Next
        Next
    End Sub

#End Region

#Region "Checkbox Dados"
    'Cambia el "enabled" de los checkers
    Private Sub EstadoCheckers(boo As Boolean)
        CBD1.Enabled = boo
        CBD2.Enabled = boo
        CBD3.Enabled = boo
        CBD4.Enabled = boo
        CBD5.Enabled = boo
    End Sub

    'Se utiliza para deschequear los dados, asi como tambien para limpiarlos
    Private Sub limpiaDados()
        CBD1.Checked = False
        CBD2.Checked = False
        CBD3.Checked = False
        CBD4.Checked = False
        CBD5.Checked = False

        EstadoCheckers(False)

        CBD1.Text = ""
        CBD2.Text = ""
        CBD3.Text = ""
        CBD4.Text = ""
        CBD5.Text = ""
    End Sub

    'actualiza los dados de un lanzamiento de un jugador
    Private Sub mostrarDados(j As Integer)
        CBD1.Text = (_GeneralaBLL.liPlayers.Item(j).liDados.Item(0).cara)
        CBD2.Text = (_GeneralaBLL.liPlayers.Item(j).liDados.Item(1).cara)
        CBD3.Text = (_GeneralaBLL.liPlayers.Item(j).liDados.Item(2).cara)
        CBD4.Text = (_GeneralaBLL.liPlayers.Item(j).liDados.Item(3).cara)
        CBD5.Text = (_GeneralaBLL.liPlayers.Item(j).liDados.Item(4).cara)
    End Sub

#End Region

#Region "Botones de asignacion de resultados"

    'Se usa para habilitar o deshabilitar los botones que ingresan los puntos
    Private Sub cambiarEstadoBA(boo As Boolean)
        BAD1.Enabled = boo
        BAD2.Enabled = boo
        BAD3.Enabled = boo
        BAD4.Enabled = boo
        BAD5.Enabled = boo
        BAD6.Enabled = boo
        BAEscalera.Enabled = boo
        BAFull.Enabled = boo
        BAPoker.Enabled = boo
        BAGenerala.Enabled = boo
        BAGeneralaD.Enabled = boo
    End Sub

    'activa los botones segun el jugador indicado
    Private Sub activarEstadoBA(jugador As Integer)
        BAD1.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(1)
        BAD2.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(2)
        BAD3.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(3)
        BAD4.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(4)
        BAD5.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(5)
        BAD6.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(6)
        BAEscalera.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(7)
        BAFull.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(8)
        BAPoker.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(9)
        BAGenerala.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(10)
        BAGeneralaD.Enabled = _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(0)
    End Sub

    'carga los valores conseguido en el juego a un jugador
    Private Sub cargarPuntaje(jugador As Integer, pos As Integer)
        _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorValor.Item(pos) = _tempPuntaje.Item(pos)
        _GeneralaBLL.liPlayers.Item(jugador).planillaBLL.planillaBE.anotadorEstado.Item(pos) = False
    End Sub

    'Simplemente agrupo acciones en un solo metodo comunes a todos los botones
    Private Sub agrupadorMedtodosBA()

        '#######################
        '### Pre-Condiciones ###
        '#######################



        '###########################
        '### Ejecucion del juego ###
        '###########################


        '########################
        '### Post-Condiciones ###
        '########################

        'Se utiliza para deschequear los dados, asi como tambien para limpiarlos
        limpiaDados()
        'Se usa para habilitar o deshabilitar los botones que ingresan los puntos
        cambiarEstadoBA(False)
        'Habilito el boton de lanzamiento de dados
        BLanzaDado.Enabled = True
        'Reinicio el numero de jugadas para el Jugador
        _GeneralaBLL.generalaBE.numeroJugada = 0
        'contiene todo lo que hay que refrescar por cada jugada
        actualizarEstadoJuego()
        'limpio los resultados temporales una ves asignados
        _tempPuntaje = Nothing
        actualizarTempPuntaje(False)
    End Sub


#Region "botones"
    Private Sub BAD1_Click(sender As Object, e As EventArgs) Handles BAD1.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 1)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAD2_Click(sender As Object, e As EventArgs) Handles BAD2.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 2)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAD3_Click(sender As Object, e As EventArgs) Handles BAD3.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 3)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAD4_Click(sender As Object, e As EventArgs) Handles BAD4.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 4)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAD5_Click(sender As Object, e As EventArgs) Handles BAD5.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 5)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAD6_Click(sender As Object, e As EventArgs) Handles BAD6.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 6)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAEscalera_Click(sender As Object, e As EventArgs) Handles BAEscalera.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 7)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAFull_Click(sender As Object, e As EventArgs) Handles BAFull.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 8)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAPoker_Click(sender As Object, e As EventArgs) Handles BAPoker.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 9)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAGenerala_Click(sender As Object, e As EventArgs) Handles BAGenerala.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 10)
        agrupadorMedtodosBA()

    End Sub

    Private Sub BAGeneralaD_Click(sender As Object, e As EventArgs) Handles BAGeneralaD.Click
        cargarPuntaje(_GeneralaBLL.generalaBE.turno, 0)
        agrupadorMedtodosBA()

    End Sub
#End Region

#End Region

#Region "boton lanzar dados"

    'lanza dados sobre los elementos no chequeados
    Private Sub ejecutarDados(j As Integer)
        If Not CBD1.Checked Then
            _GeneralaBLL.liPlayers.Item(j).liDados.Item(0) = _GeneralaBLL.dadoBLL.lanzarDado
        End If
        If Not CBD2.Checked Then
            _GeneralaBLL.liPlayers.Item(j).liDados.Item(1) = _GeneralaBLL.dadoBLL.lanzarDado
        End If
        If Not CBD3.Checked Then
            _GeneralaBLL.liPlayers.Item(j).liDados.Item(2) = _GeneralaBLL.dadoBLL.lanzarDado
        End If
        If Not CBD4.Checked Then
            _GeneralaBLL.liPlayers.Item(j).liDados.Item(3) = _GeneralaBLL.dadoBLL.lanzarDado
        End If
        If Not CBD5.Checked Then
            _GeneralaBLL.liPlayers.Item(j).liDados.Item(4) = _GeneralaBLL.dadoBLL.lanzarDado
        End If




    End Sub

    'Boton que se encarga de lanzar los dados y actualizar algunos estados.
    Private Sub BLanzaDado_Click(sender As Object, e As EventArgs) Handles BLanzaDado.Click

        'guardo temporalmente turno (jugador correspondiente) para una escritura mas legible
        Dim jugador As Integer = _GeneralaBLL.generalaBE.turno

        '#######################
        '### Pre-Condiciones ###
        '#######################

        'limpio el puntaje para iniciar una nueva carga
        _tempPuntaje = Nothing
        'activa los botones segun el jugador indicado
        activarEstadoBA(jugador)
        'Cambia el "enabled" de los checkers
        EstadoCheckers(True)

        '###########################
        '### Ejecucion del juego ###
        '###########################

        'lanza dados sobre los elementos no chequeados
        ejecutarDados(jugador)
        'actualiza los dados de un lanzamiento de un jugador
        mostrarDados(jugador)

        'guardo el puntaje temporal segun cada tirada de dados.
        _tempPuntaje = _GeneralaBLL.chequear(jugador, _GeneralaBLL.generalaBE.numeroJugada)
        'Actualiza los labels correspondientes al puntaje temporario
        actualizarTempPuntaje(True)

        '########################
        '### Post-Condiciones ###
        '########################

        'cuenta la cantidad de tiradas de un jugador
        _GeneralaBLL.TiradaNext()
        'Muestra la jugada en curso
        LNumeroTirada.Text = _GeneralaBLL.generalaBE.numeroJugada

        'deshabilita el boton de lanzar dados una vez superado los 3 lanzamientos.
        If _GeneralaBLL.generalaBE.numeroJugada >= 3 Then
            BLanzaDado.Enabled = False
            'Cambia el "enabled" de los checkers
        End If

    End Sub

#End Region

#Region "handle de eventos"

    'capta el evento cuando se produce una generala servida y registra al jugador
    Private Sub vicGS(ByVal jugador As BE.USUARIO) Handles _GeneralaBLL.generalaServida
        ganador(jugador)
    End Sub

    'cuando se termina un juego, capta el evento, busca a quien tiene mayor puntaje y lo registra.
    Private Sub vicPuntos() Handles _GeneralaBLL.finJuego
        'hago una busqueda del jugador que mas puntos sumo
        Dim puntos As Integer = -1
        Dim jugador As BE.USUARIO = Nothing
        For Each i In _GeneralaBLL.liPlayers
            If i.planillaBLL.resultado > puntos Then
                puntos = i.planillaBLL.resultado
                jugador = i.usuarioBE
            End If
        Next
        ganador(jugador)
    End Sub

    'contiene todas la rutina en comun para los eventos de victoria
    Private Sub ganador(usr As BE.USUARIO)
        'deshabilito el tablero para obligarlo a generar una nueva partida si quiere seguir jugando.
        Me.Enabled = False

        'muestra un cartel de quien fue el ganador
        MsgBox("El jugador " & usr.usuario & " ha ganado el juego!")
        'actualizo las victolias en la lista de jugadores logueados
        '(para no tener que volver a buscar en la DB
        For Each jug In FMCONTENEDOR.liUSRLogin
            If jug.usuario = usr.usuario Then
                jug.victorias += 1
                'guarda en la DB la victoria del jugador
                _GeneralaBLL.guardarVictoria(usr)
            End If
        Next

        For Each jug In FMCONTENEDOR.liUSRPartida
            For Each jl In FMCONTENEDOR.liUSRLogin
                If jug.usuario = jl.usuario Then
                    jl.partidos += 1
                    _GeneralaBLL.guardarPartidos(jl)
                End If
            Next
        Next
    End Sub

#End Region

End Class

File: GUI/FMTABLERO.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace GUI
{
    public partial class FMTABLERO : Form
    {
        public FMTABLERO()
        {
            InitializeComponent();
        }


        //variable sobre la cual el formulario guarda el juego para operar
        BLL.GENERALA _generalaBLL;

        //lista de puntajes temporales posibles de un jugador dentro de la tirada para luego ser asignada
        List<int> _tempPuntaje;

        //Carga el formulario del Tablero.
        private void FMTABLERO_Load(object sender, EventArgs e)
        {
            //cargo el juego nuevo
            _generalaBLL = new BLL.GENERALA();

            MEH manejadorEventosHandle = new MEH(_generalaBLL, this);

            _generalaBLL.generalaServidaEvento += manejadorEventosHandle.vicGS;
            _generalaBLL.finJuegoEvento += manejadorEventosHandle.vicPuntos;

            //tomo los usuarios de una nueva partida y los coloco en una lista de PLAYERS
            List<BLL.PLAYER> PL = new List<BLL.PLAYER>();
            foreach (var i in FMCONTENEDOR.liUSRPartida)
	        {
                PL.Add(new BLL.PLAYER() { usuarioBE = i });
	        }
            //le agrego la lista de players al juego
            _generalaBLL.agregarPlayers(PL);

            //pongo en 0 todos los datos de la GUI
            actualizarEstadoJuego();
            limpiaDados();
            cambiarEstadoBA(false);
            actualizarTempPuntaje(false);

        }

        #region Actualizar Labels

        //contiene todo lo que hay que refrescar por cada jugada
        private void actualizarEstadoJuego()
        {
            //genera el siguiente turno
            _generalaBLL.turnoNext();
            //muestra la tirada en curso
            LNumeroJugada.Text = _generalaBLL.GeneralaBE.rondas.ToString();

            //actualiza label Jugador
            LTurnoJugador.Text = _generalaBLL.LiPlayers[_generalaBLL.GeneralaBE.turno].usuarioBE.usuario;
            //actualiza el label de numero de tirada
            LNumeroTirada.Text = _generalaBLL.GeneralaBE.numeroJugada.ToString();
            //actualiza la tabla de puntajes para cada jugador
            for (int i = 0; i < _generalaBLL.LiPlayers.Count; i++)
            {
                refreshScore(i);
            }


        }

        private void refreshScore(int jugador)
        {
            string lb1, lb2;
            //aca guardo el control a buscar
            for (int i = 0; i < 11; i++)
            {
                lb1 = string.Format("SJ {0} A {1}", jugador + 1, i);
                foreach (Control c in this.Controls)
                {
                    if (c.Name == lb1)
                    {
                        c.Text = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorValores[i].ToString();
                    }
                }
            }
            //armo un texto con el nombre del "label" que voy a buscar
            lb1 = string.Format("SJ {0} Resultado", jugador + 1);
            lb2 = string.Format("LJugador {0}", jugador + 1);
            //recorro todos los controles para encontrar el control del "label" que quiero buscar
            foreach (Control c in this.Controls)
            {
                //chequeo el nombre del "label" que busco
                if (c.Name == lb1)
                {
                    //asigno el contenido de informacion que necesito al "label" encontrado
                    c.Text = _generalaBLL.LiPlayers[jugador].planillaBLL.resultado().ToString();
                }
                if (c.Name == lb2)
                {
                    //asigno el contenido de informacion que necesito al "label" encontrado
                    c.Text = _generalaBLL.LiPlayers[jugador].usuarioBE.usuario;
                }
            }
        }

        //Actualiza los labels correspondientes al puntaje temporario
        private void actualizarTempPuntaje(bool boo)
        {
            string lb;
            for (int i = 0; i < 11; i++)
            {
                lb = string.Format("LST {0}", i);
                foreach (Control c in this.Controls)
                {
                    if (c.Name == lb)
                    {
                        if (boo)
                        {
                            c.Text = _tempPuntaje[i].ToString();
                        }
                        else
                        {
                            c.Text = "";
                        }
                    }
                }
            }
        }

        #endregion Actualizar Labels

        #region Checkbox Dados

        //Cambia el "enabled" de los checkers
        private void EstadoCheckers(bool boo)
        {
            CBD1.Enabled = boo;
            CBD2.Enabled = boo;
            CBD3.Enabled = boo;
            CBD4.Enabled = boo;
            CBD5.Enabled = boo;
        }

        //Se utiliza para deschequear los dados, asi como tambien para limpiarlos
        private void limpiaDados()
        {
            CBD1.Checked = false;
            CBD2.Checked = false;
            CBD3.Checked = false;
            CBD4.Checked = false;
            CBD5.Checked = false;

            EstadoCheckers(false);

            CBD1.Text = "";
            CBD2.Text = "";
            CBD3.Text = "";
            CBD4.Text = "";
            CBD5.Text = "";
        }

        //actualiza los dados de un lanzamiento de un jugador
        private void mostrarDados(int j)
        {
            CBD1.Text = _generalaBLL.LiPlayers[j].liDados[0].cara.ToString();
            CBD2.Text = _generalaBLL.LiPlayers[j].liDados[1].cara.ToString();
            CBD3.Text = _generalaBLL.LiPlayers[j].liDados[2].cara.ToString();
            CBD4.Text = _generalaBLL.LiPlayers[j].liDados[3].cara.ToString();
            CBD5.Text = _generalaBLL.LiPlayers[j].liDados[4].cara.ToString();
        }

        #endregion Checkbox Dados

        #region Botones de asignacion de resultados

        //Se usa para habilitar o deshabilitar los botones que ingresan los puntos
        private void cambiarEstadoBA(bool boo)
        {
            BAD1.Enabled = boo;
            BAD2.Enabled = boo;
            BAD3.Enabled = boo;
            BAD4.Enabled = boo;
            BAD5.Enabled = boo;
            BAD6.Enabled = boo;
            BAEscalera.Enabled = boo;
            BAFull.Enabled = boo;
            BAPoker.Enabled = boo;
            BAGenerala.Enabled = boo;
            BAGeneralaD.Enabled = boo;
        }

        //activa los botones segun el jugador indicado
        private void activarEstadoBA(int jugador)
        {
            BAD1.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[1];
            BAD2.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[2];
            BAD3.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[3];
            BAD4.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[4];
            BAD5.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[5];
            BAD6.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[6];
            BAEscalera.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[7];
            BAFull.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[8];
            BAPoker.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[9];
            BAGenerala.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[10];
            BAGeneralaD.Enabled = _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[0];
        }

        //carga los valores conseguido en el juego a un jugador
        private void cargarPuntaje(int jugador, int pos)
        {
            _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorValores[pos] = _tempPuntaje[pos];
            _generalaBLL.LiPlayers[jugador].planillaBLL.planillaBE.AnotadorEstado[pos] = false;
        }

        //Simplemente agrupo acciones en un solo metodo comunes a todos los botones
        private void agrupadorMetodosBA()
        {
            //#######################
            //### Pre-Condiciones ###
            //#######################
            
            //###########################
            //### Ejecucion del juego ###
            //###########################
            
            //########################
            //### Post-Condiciones ###
            //########################

            //Se utiliza para deschequear los dados, asi como tambien para limpiarlos
            limpiaDados();
            //Se usa para habilitar o deshabilitar los botones que ingresan los puntos
            cambiarEstadoBA(false);
            //Habilito el boton de lanzamiento de dados
            BLanzaDado.Enabled = true;
            //Reinicio el numero de jugadas para el Jugador
            _generalaBLL.GeneralaBE.numeroJugada = 0;
            //contiene todo lo que hay que refrescar por cada jugada
            actualizarEstadoJuego();
            //limpio los resultados temporales una ves asignados
            _tempPuntaje = null;
            actualizarTempPuntaje(false);
        }

        #region Botones

        private void BAD1_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 1);
            agrupadorMetodosBA();
        }

        private void BAD2_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 2);
            agrupadorMetodosBA();
        }

        private void BAD3_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 3);
            agrupadorMetodosBA();
        }

        private void BAD4_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 4);
            agrupadorMetodosBA();
        }

        private void BAD5_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 5);
            agrupadorMetodosBA();
        }

        private void BAD6_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 6);
            agrupadorMetodosBA();
        }

        private void BAEscalera_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 7);
            agrupadorMetodosBA();
        }

        private void BAFull_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 8);
            agrupadorMetodosBA();
        }

        private void BAPoker_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 9);
            agrupadorMetodosBA();
        }

        private void BAGenerala_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 10);
            agrupadorMetodosBA();
        }

        private void BAGeneralaD_Click(object sender, EventArgs e)
        {
            cargarPuntaje(_generalaBLL.GeneralaBE.turno, 0);
            agrupadorMetodosBA();
        }

        #endregion Botones

        #endregion Botones de asignacion de resultados

        #region Boton lanzar dados

        //lanza dados sobre los elementos no chequeados
        private void ejecutarDados(int j)
        {
            if (!CBD1.Checked)
            {
                _generalaBLL.LiPlayers[j].liDados[0] = _generalaBLL.DadoBLL.lanzarDado();
            }
            if (!CBD2.Checked)
            {
                _generalaBLL.LiPlayers[j].liDados[1] = _generalaBLL.DadoBLL.lanzarDado();
            }
            if (!CBD3.Checked)
            {
                _generalaBLL.LiPlayers[j].liDados[2] = _generalaBLL.DadoBLL.lanzarDado();
            }
            if (!CBD4.Checked)
            {
                _generalaBLL.LiPlayers[j].liDados[3] = _generalaBLL.DadoBLL.lanzarDado();
            }
            if (!CBD5.Checked)
            {
                _generalaBLL.LiPlayers[j].liDados[4] = _generalaBLL.DadoBLL.lanzarDado();
            }
        }

        ////Boton que se encarga de lanzar los dados y actualizar algunos estados.
        private void BLanzaDado_Click(object sender, EventArgs e)
        {
            //guardo temporalmente turno (jugador correspondiente) para una escritura mas legible
            int jugador = _generalaBLL.GeneralaBE.turno;

            //#######################
            //### Pre-Condiciones ###
            //#######################

            //limpio el puntaje para iniciar una nueva carga
            _tempPuntaje = null;
            //activa los botones segun el jugador indicado
            activarEstadoBA(jugador);
            //Cambia el "enabled" de los checkers
            EstadoCheckers(true);

            //###########################
            //### Ejecucion del juego ###
            //###########################

            //lanza dados sobre los elementos no chequeados
            ejecutarDados(jugador);
            //actualiza los dados de un lanzamiento de un jugador
            mostrarDados(jugador);

            //guardo el puntaje temporal segun cada tirada de dados.
            _tempPuntaje = _generalaBLL.chequear(jugador, _generalaBLL.GeneralaBE.numeroJugada);
            //Actualiza los labels correspondientes al puntaje temporario
            actualizarTempPuntaje(true);

            //########################
            //### Post-Condiciones ###
            //########################

            //cuenta la cantidad de tiradas de un jugador
            _generalaBLL.tiradaNext();
            //Muestra la jugada en curso
            LNumeroTirada.Text = _generalaBLL.GeneralaBE.numeroJugada.ToString();

            //deshabilita el boton de lanzar dados una vez superado los 3 lanzamientos.
            if (_generalaBLL.GeneralaBE.numeroJugada >= 3)
            {
                //Cambia el "enabled" de los checkers
                BLanzaDado.Enabled = false;
            }
            
        }

        #endregion Boton lanzar dados

        #region Eventos

        public class MEH
        {
            private BLL.GENERALA _g;
            private FMTABLERO _me;

            public MEH(BLL.GENERALA generalaBLL, FMTABLERO me)
            {
                _g = generalaBLL;
                _me = me;
            }

            //capta el evento cuando se produce una generala servida y registra al jugador
            public void vicGS(BE.USUARIO jugador)
            {
                ganador(jugador);
            }

            //cuando se termina un juego, capta el evento, busca a quien tiene mayor puntaje y lo registra.
            public void vicPuntos()
            {
                //hago una busqueda del jugador que mas puntos sumo
                int puntos = -1;
                BE.USUARIO jugador = null;
                foreach (BLL.PLAYER i in _g.LiPlayers)
                {
                    if (i.planillaBLL.resultado() > puntos)
                    {
                        puntos = i.planillaBLL.resultado();
                        jugador = i.usuarioBE;
                    }
                }
                ganador(jugador);
            }

            //contiene todas la rutina en comun para los eventos de victoria
            public void ganador(BE.USUARIO usr)
            {
                //deshabilito el tablero para obligarlo a generar una nueva partida si quiere seguir jugando.
                _me.Enabled = false;

                //muestra un cartel de quien fue el ganador
                MessageBox.Show("El jugador {0} ha ganado el juego!", usr.usuario);
                //actualizo las victolias en la lista de jugadores logueados
                //(para no tener que volver a buscar en la DB
                foreach (BE.USUARIO jug in FMCONTENEDOR.liUSRLogin)
                {
                    if (jug.usuario == usr.usuario)
                    {
                        jug.victorias += 1;
                        //guarda en la DB la victoria del jugador
                        _g.guardarVictoria(usr);
                    }
                }
                //busco a los jugadores de la partida entre los logueados y registro el partido jugado
                foreach (BE.USUARIO jug in FMCONTENEDOR.liUSRPartida)
                {
                    foreach (BE.USUARIO jl in FMCONTENEDOR.liUSRLogin)
                    {
                        if (jug.usuario == jl.usuario)
                        {
                            jl.partidos += 1;
                            _g.guardarPartidos(jl);
                        }
                    }
                }

            }
        }

        #endregion Eventos

    }


    

}
